File: flutter/lib/src/chat_session.dart
/// Multi-turn conversation session management
///
/// [ChatSession] wraps an [EdgeVeda] instance and manages conversation
/// history automatically. Developers call [send] or [sendStream] with
/// a user prompt, and ChatSession handles:
/// - Formatting the full conversation history using the model's chat template
/// - Tracking user/assistant messages
/// - Summarizing older messages when the context window fills up
///
/// System prompts are set at creation time and remain immutable for the
/// session's lifetime. Use [reset] to start a fresh conversation while
/// keeping the model loaded.
///
/// Example:
/// ```dart
/// final session = ChatSession(
///   edgeVeda: edgeVeda,
///   preset: SystemPromptPreset.assistant,
/// );
///
/// // Full response
/// final reply = await session.send('What is Flutter?');
/// print(reply.content);
///
/// // Streaming response
/// await for (final chunk in session.sendStream('Tell me more')) {
///   stdout.write(chunk.token);
/// }
///
/// // Check context usage
/// print('Turns: ${session.turnCount}, Context: ${(session.contextUsage * 100).toInt()}%');
///
/// // Start fresh conversation (model stays loaded)
/// session.reset();
/// ```
library;

import 'chat_template.dart';
import 'chat_types.dart';
import 'edge_veda_impl.dart';
import 'types.dart'
    show
        CancelToken,
        ConfigurationException,
        GenerateOptions,
        TokenChunk;

/// Manages multi-turn conversation state on top of [EdgeVeda]
///
/// ChatSession is pure Dart -- it uses no new C API symbols. It formats
/// conversation history using chat templates and delegates inference to
/// the existing [EdgeVeda.generate] and [EdgeVeda.generateStream] methods.
class ChatSession {
  final EdgeVeda _edgeVeda;

  /// The system prompt for this session (immutable after creation)
  ///
  /// Set via constructor parameter or [SystemPromptPreset]. If both
  /// [systemPrompt] and [preset] are provided, [systemPrompt] takes
  /// precedence.
  final String? systemPrompt;

  /// The chat template format used for formatting prompts
  final ChatTemplateFormat templateFormat;

  final int _contextLength;
  final int _maxResponseTokens;
  final List<ChatMessage> _messages = [];
  bool _isSummarizing = false;

  /// Create a new chat session
  ///
  /// Requires an initialized [EdgeVeda] instance. Throws
  /// [ConfigurationException] if the instance is not initialized.
  ///
  /// [systemPrompt] sets a custom system prompt. If null and [preset]
  /// is provided, the preset's prompt text is used instead.
  ///
  /// [templateFormat] defaults to [ChatTemplateFormat.llama3Instruct]
  /// for Llama 3.x models. Change this if using a different model family.
  ///
  /// [maxResponseTokens] reserves space in the context window for the
  /// model's response (defaults to 512 tokens).
  ChatSession({
    required EdgeVeda edgeVeda,
    String? systemPrompt,
    SystemPromptPreset? preset,
    this.templateFormat = ChatTemplateFormat.llama3Instruct,
    int maxResponseTokens = 512,
  })  : _edgeVeda = edgeVeda,
        systemPrompt = systemPrompt ?? preset?.prompt,
        _contextLength = edgeVeda.config?.contextLength ?? 2048,
        _maxResponseTokens = maxResponseTokens {
    if (!edgeVeda.isInitialized) {
      throw const ConfigurationException(
        'EdgeVeda must be initialized before creating a ChatSession. Call init() first.',
      );
    }
  }

  /// Read-only access to conversation history
  ///
  /// Returns an unmodifiable view of the message list. Messages are in
  /// chronological order. Includes user messages, assistant responses,
  /// and any summary messages from context overflow handling.
  List<ChatMessage> get messages => List.unmodifiable(_messages);

  /// Number of user turns in the conversation
  int get turnCount => _messages.where((m) => m.role == ChatRole.user).length;

  /// Estimated context window usage as a fraction (0.0 to 1.0+)
  ///
  /// Uses a rough heuristic of ~4 characters per token. This is an
  /// approximation -- exact token counting requires the model's vocabulary
  /// which is not exposed via the current API.
  ///
  /// Values above 0.7 may trigger automatic summarization on the next
  /// [send] or [sendStream] call.
  double get contextUsage {
    if (_contextLength <= 0) return 0.0;
    final formatted = _formatConversation();
    final estimatedTokens = formatted.length ~/ 4;
    return estimatedTokens / _contextLength;
  }

  /// Whether a summarization is currently in progress
  bool get isSummarizing => _isSummarizing;

  /// Send a message and get the complete response
  ///
  /// Adds the user message to history, checks for context overflow
  /// (triggering summarization if needed), formats the conversation,
  /// generates a response, and adds the assistant reply to history.
  ///
  /// Returns the assistant's [ChatMessage] with the complete response.
  ///
  /// On error, the user message is rolled back from history to keep
  /// the conversation in a consistent state.
  ///
  /// Example:
  /// ```dart
  /// final reply = await session.send('What is Dart?');
  /// print(reply.content);
  /// print('Turn ${session.turnCount}');
  /// ```
  Future<ChatMessage> send(
    String prompt, {
    GenerateOptions? options,
    CancelToken? cancelToken,
  }) async {
    // Add user message
    _messages.add(ChatMessage(
      role: ChatRole.user,
      content: prompt,
      timestamp: DateTime.now(),
    ));

    try {
      // Check and summarize if needed
      await _summarizeIfNeeded(cancelToken: cancelToken);

      // Format full conversation
      final formatted = _formatConversation();

      // Generate response
      final response = await _edgeVeda.generate(formatted, options: options);

      // Add assistant message
      final assistantMsg = ChatMessage(
        role: ChatRole.assistant,
        content: response.text,
        timestamp: DateTime.now(),
      );
      _messages.add(assistantMsg);
      return assistantMsg;
    } catch (e) {
      // Rollback user message on error
      if (_messages.isNotEmpty && _messages.last.role == ChatRole.user) {
        _messages.removeLast();
      }
      rethrow;
    }
  }

  /// Send a message and stream the response token-by-token
  ///
  /// Adds the user message to history, checks for context overflow,
  /// formats the conversation, and streams the model's response. Each
  /// [TokenChunk] contains a token fragment. After the stream completes,
  /// the full assistant response is added to history.
  ///
  /// On error, the user message is rolled back from history.
  ///
  /// Example:
  /// ```dart
  /// final buffer = StringBuffer();
  /// await for (final chunk in session.sendStream('Tell me a joke')) {
  ///   if (!chunk.isFinal) {
  ///     buffer.write(chunk.token);
  ///     stdout.write(chunk.token);
  ///   }
  /// }
  /// print('\nFull response: $buffer');
  /// ```
  Stream<TokenChunk> sendStream(
    String prompt, {
    GenerateOptions? options,
    CancelToken? cancelToken,
  }) async* {
    // Add user message
    _messages.add(ChatMessage(
      role: ChatRole.user,
      content: prompt,
      timestamp: DateTime.now(),
    ));

    try {
      // Check and summarize if needed
      await _summarizeIfNeeded(cancelToken: cancelToken);

      // Format full conversation
      final formatted = _formatConversation();

      // Stream response, collecting tokens
      final buffer = StringBuffer();
      await for (final chunk in _edgeVeda.generateStream(
        formatted,
        options: options,
        cancelToken: cancelToken,
      )) {
        if (!chunk.isFinal) {
          buffer.write(chunk.token);
        }
        yield chunk;
      }

      // Add complete assistant message to history
      final responseText = buffer.toString();
      if (responseText.isNotEmpty) {
        _messages.add(ChatMessage(
          role: ChatRole.assistant,
          content: responseText,
          timestamp: DateTime.now(),
        ));
      }
    } catch (e) {
      // Rollback user message on error
      if (_messages.isNotEmpty && _messages.last.role == ChatRole.user) {
        _messages.removeLast();
      }
      rethrow;
    }
  }

  /// Reset conversation history (keep model loaded)
  ///
  /// Clears all messages but preserves the system prompt and model state.
  /// The next [send] or [sendStream] call starts a fresh conversation
  /// with fast response time (no model reload needed).
  void reset() {
    _messages.clear();
  }

  /// Format the current conversation into a prompt string
  String _formatConversation() {
    return ChatTemplate.format(
      template: templateFormat,
      systemPrompt: systemPrompt,
      messages: _messages,
    );
  }

  /// Check if context window is getting full and summarize if needed
  ///
  /// Triggers summarization when estimated token usage exceeds 70% of
  /// available capacity (context length minus reserved response tokens).
  ///
  /// Keeps the last 2 user turns and their assistant replies intact.
  /// Older messages are summarized by the model and replaced with a
  /// single summary message.
  ///
  /// If summarization fails, falls back to simple truncation (dropping
  /// oldest messages until within budget). Never crashes.
  Future<void> _summarizeIfNeeded({CancelToken? cancelToken}) async {
    final formatted = _formatConversation();
    final estimatedTokens = formatted.length ~/ 4;
    final availableTokens = _contextLength - _maxResponseTokens;

    // Only summarize when above 70% of available capacity
    if (estimatedTokens < availableTokens * 0.7) return;

    _isSummarizing = true;
    try {
      // Find split point: keep last 2 user turns + their assistant replies
      int userCount = 0;
      int splitIndex = _messages.length;
      for (int i = _messages.length - 1; i >= 0; i--) {
        if (_messages[i].role == ChatRole.user) {
          userCount++;
        }
        if (userCount >= 2) {
          splitIndex = i;
          break;
        }
      }

      // Nothing to summarize if split point is at or before start
      if (splitIndex <= 0) return;

      // Extract old and recent messages
      final oldMessages = _messages.sublist(0, splitIndex);
      final recentMessages = _messages.sublist(splitIndex);

      // Build summarization prompt
      final summaryPrompt = StringBuffer();
      summaryPrompt.writeln(
          'Summarize this conversation concisely. Keep key facts and decisions:');
      for (final msg in oldMessages) {
        if (msg.role == ChatRole.summary) {
          summaryPrompt.writeln('summary: ${msg.content}');
        } else {
          summaryPrompt.writeln('${msg.role.name}: ${msg.content}');
        }
      }

      // Generate summary using the model (low temperature for factual output)
      final summaryResponse = await _edgeVeda.generate(
        summaryPrompt.toString(),
        options: const GenerateOptions(maxTokens: 128, temperature: 0.3),
      );

      // Replace old messages with summary + recent messages
      _messages.clear();
      _messages.add(ChatMessage(
        role: ChatRole.summary,
        content: summaryResponse.text,
        timestamp: DateTime.now(),
      ));
      _messages.addAll(recentMessages);
    } catch (e) {
      // Fallback: simple truncation if summarization fails
      // Drop oldest messages until estimated tokens are under 60% of available
      final availableTokens = _contextLength - _maxResponseTokens;
      final targetTokens = (availableTokens * 0.6).toInt();

      while (_messages.length > 2) {
        final currentFormatted = _formatConversation();
        final currentTokens = currentFormatted.length ~/ 4;
        if (currentTokens <= targetTokens) break;
        _messages.removeAt(0);
      }

      // Log warning but never crash
      print(
          'ChatSession: Summarization failed, fell back to truncation. Error: $e');
    } finally {
      _isSummarizing = false;
    }
  }
}

File: react-native/src/ChatSession.ts
import EdgeVedaInstance from './EdgeVeda';
import { ChatMessage, ChatRole, SystemPromptPreset, getSystemPrompt } from './ChatTypes';
import { ChatTemplate, formatMessages } from './ChatTemplate';
import { ToolRegistry, ToolDefinition, ToolCall, ToolResult } from './ToolRegistry';
import { ToolTemplate, ToolTemplateFormat } from './ToolTemplate';
import { GbnfBuilder } from './GbnfBuilder';
import { SchemaValidator } from './SchemaValidator';
import { CancelToken, GenerateOptions, GenerationError, TokenCallback } from './types';

/**
 * EdgeVeda SDK instance type
 */
type EdgeVeda = typeof EdgeVedaInstance;

/**
 * Manages a multi-turn chat conversation with context tracking.
 *
 * ChatSession wraps an EdgeVeda instance and manages conversation history
 * automatically. Supports plain generation, tool calling, and structured
 * (grammar-constrained) JSON output. Mirrors Flutter SDK's ChatSession.
 *
 * Context overflow is handled by automatic summarization — when the
 * estimated token count exceeds 70% of the context window, older messages
 * are compressed into a summary to keep inference within budget.
 */
export class ChatSession {
  private edgeVeda: EdgeVeda;
  private messages: ChatMessage[] = [];
  private maxContextLength: number;
  private maxResponseTokens: number;
  private template: ChatTemplate;
  private systemPromptText: string | null;
  private _tools: ToolRegistry | null;
  private _toolFormat: ToolTemplateFormat;
  private _isSummarizing = false;

  /**
   * Creates a new chat session.
   *
   * @param edgeVeda - Initialized EdgeVeda instance
   * @param options.systemPrompt - Pre-configured system prompt preset
   * @param options.systemPromptText - Custom system prompt string (overrides preset)
   * @param options.maxContextLength - Context window size (default 2048)
   * @param options.maxResponseTokens - Tokens reserved for model reply (default 512)
   * @param options.template - Chat template format (default LLAMA3)
   * @param options.tools - ToolRegistry for function calling support
   * @param options.toolFormat - Tool call format (default derived from template)
   */
  constructor(
    edgeVeda: EdgeVeda,
    options: {
      systemPrompt?: SystemPromptPreset;
      systemPromptText?: string;
      maxContextLength?: number;
      maxResponseTokens?: number;
      template?: ChatTemplate;
      tools?: ToolRegistry;
      toolFormat?: ToolTemplateFormat;
    } = {}
  ) {
    this.edgeVeda = edgeVeda;
    this.maxContextLength = options.maxContextLength ?? 2048;
    this.maxResponseTokens = options.maxResponseTokens ?? 512;
    this.template = options.template ?? ChatTemplate.LLAMA3;
    this._tools = options.tools ?? null;
    this._toolFormat = options.toolFormat ?? this._deriveToolFormat();

    // Resolve system prompt: explicit string takes priority over preset
    if (options.systemPromptText) {
      this.systemPromptText = options.systemPromptText;
    } else if (options.systemPrompt) {
      this.systemPromptText = getSystemPrompt(options.systemPrompt);
    } else {
      this.systemPromptText = null;
    }

    if (this.systemPromptText) {
      this.messages.push({
        role: ChatRole.SYSTEM,
        content: this.systemPromptText,
        timestamp: new Date(),
      });
    }
  }

  // ---------------------------------------------------------------------------
  // Public getters
  // ---------------------------------------------------------------------------

  /** Whether a context summarization is currently in progress */
  get isSummarizing(): boolean {
    return this._isSummarizing;
  }

  /** The tool registry for this session, or null if no tools registered */
  get toolRegistry(): ToolRegistry | null {
    return this._tools;
  }

  /** Number of user turns in the conversation */
  get turnCount(): number {
    return this.messages.filter((m) => m.role === ChatRole.USER).length;
  }

  /**
   * Estimated context window usage as a fraction (0.0 to 1.0+).
   *
   * Uses a rough heuristic of ~4 characters per token.
   * Values above 0.7 trigger automatic summarization on the next send().
   */
  get contextUsage(): number {
    const totalTokens = this.estimateTokens();
    return totalTokens / this.maxContextLength;
  }

  /** All messages in the conversation (read-only copy) */
  get allMessages(): ChatMessage[] {
    return [...this.messages];
  }

  /** Returns the last N messages in the conversation */
  lastMessages(count: number): ChatMessage[] {
    const startIndex = Math.max(0, this.messages.length - count);
    return this.messages.slice(startIndex);
  }

  // ---------------------------------------------------------------------------
  // Core send methods
  // ---------------------------------------------------------------------------

  /**
   * Sends a message and returns the complete response.
   *
   * Adds the user message to history, checks for context overflow
   * (triggering summarization if needed), generates a response, and
   * adds the assistant reply to history.
   *
   * On error, the user message is rolled back to keep conversation consistent.
   */
  async send(
    message: string,
    options?: GenerateOptions,
    cancelToken?: CancelToken
  ): Promise<string> {
    cancelToken?.throwIfCancelled();

    this.messages.push({
      role: ChatRole.USER,
      content: message,
      timestamp: new Date(),
    });

    try {
      await this._summarizeIfNeeded(cancelToken);
      cancelToken?.throwIfCancelled();

      const prompt = this.formatPrompt();
      const responseText = await this.edgeVeda.generate(prompt, options);

      this.messages.push({
        role: ChatRole.ASSISTANT,
        content: responseText,
        timestamp: new Date(),
      });

      return responseText;
    } catch (e) {
      // Roll back user message on error
      if (
        this.messages.length > 0 &&
        this.messages[this.messages.length - 1].role === ChatRole.USER
      ) {
        this.messages.pop();
      }
      throw e;
    }
  }

  /**
   * Sends a message and streams the response token by token.
   *
   * On error, the user message is rolled back from history.
   */
  async sendStream(
    message: string,
    onToken: TokenCallback,
    options?: GenerateOptions,
    cancelToken?: CancelToken
  ): Promise<void> {
    cancelToken?.throwIfCancelled();

    this.messages.push({
      role: ChatRole.USER,
      content: message,
      timestamp: new Date(),
    });

    try {
      await this._summarizeIfNeeded(cancelToken);
      cancelToken?.throwIfCancelled();

      const prompt = this.formatPrompt();
      let fullResponse = '';

      await this.edgeVeda.generateStream(
        prompt,
        (token, done) => {
          if (cancelToken?.isCancelled) return;
          fullResponse += token;
          onToken(token, done);

          if (done) {
            this.messages.push({
              role: ChatRole.ASSISTANT,
              content: fullResponse,
              timestamp: new Date(),
            });
          }
        },
        options
      );
    } catch (e) {
      // Roll back user message on error
      if (
        this.messages.length > 0 &&
        this.messages[this.messages.length - 1].role === ChatRole.USER
      ) {
        this.messages.pop();
      }
      throw e;
    }
  }

  // ---------------------------------------------------------------------------
  // Tool calling
  // ---------------------------------------------------------------------------

  /**
   * Send a message with tool calling support.
   *
   * If the model responds with a tool call, `onToolCall` is invoked.
   * The developer executes the tool and returns a ToolResult. The result
   * is injected back into the conversation for a follow-up generation.
   *
   * If the model responds with plain text (no tool call detected),
   * it is returned as a normal assistant message.
   *
   * @param message - User message text
   * @param onToolCall - Async callback invoked for each tool call
   * @param options - Generation options
   * @param cancelToken - Optional cancellation token
   * @param maxToolRounds - Max tool call iterations (default 3)
   * @returns Final assistant response text
   */
  async sendWithTools(
    message: string,
    onToolCall: (toolCall: ToolCall) => Promise<ToolResult>,
    options?: GenerateOptions,
    cancelToken?: CancelToken,
    maxToolRounds: number = 3
  ): Promise<string> {
    cancelToken?.throwIfCancelled();

    this.messages.push({
      role: ChatRole.USER,
      content: message,
      timestamp: new Date(),
    });

    try {
      await this._summarizeIfNeeded(cancelToken);

      for (let round = 0; round < maxToolRounds; round++) {
        cancelToken?.throwIfCancelled();

        const prompt = this._formatConversationWithTools();
        const responseText = await this.edgeVeda.generate(prompt, options);

        // Check if the response contains a tool call
        const toolCalls = ToolTemplate.parseToolCalls(this._toolFormat, responseText);

        if (!toolCalls || toolCalls.length === 0) {
          // No tool call — plain assistant response
          this.messages.push({
            role: ChatRole.ASSISTANT,
            content: responseText,
            timestamp: new Date(),
          });
          return responseText;
        }

        // Process the first tool call (single-call per round)
        const toolCall = toolCalls[0];

        // Record the tool call in history
        this.messages.push({
          role: ChatRole.TOOL_CALL,
          content: JSON.stringify({ name: toolCall.name, arguments: toolCall.arguments }),
          timestamp: new Date(),
        });

        // Invoke the developer's handler
        const toolResult = await onToolCall(toolCall);

        // Record the tool result in history
        this.messages.push({
          role: ChatRole.TOOL_RESULT,
          content: toolResult.isError
            ? JSON.stringify({ error: toolResult.error })
            : JSON.stringify(toolResult.data),
          timestamp: new Date(),
        });
      }

      // Max rounds exhausted — generate a final response without tool parsing
      cancelToken?.throwIfCancelled();
      const finalPrompt = this._formatConversationWithTools();
      const finalText = await this.edgeVeda.generate(finalPrompt, options);

      this.messages.push({
        role: ChatRole.ASSISTANT,
        content: finalText,
        timestamp: new Date(),
      });

      return finalText;
    } catch (e) {
      // Roll back user message on error
      if (
        this.messages.length > 0 &&
        this.messages[this.messages.length - 1].role === ChatRole.USER
      ) {
        this.messages.pop();
      }
      throw e;
    }
  }

  // ---------------------------------------------------------------------------
  // Structured (grammar-constrained) generation
  // ---------------------------------------------------------------------------

  /**
   * Send a message and get a structured JSON response validated against a schema.
   *
   * Uses GBNF grammar-constrained decoding to guarantee the output is valid
   * JSON conforming to `schema`. The response is validated before delivery.
   *
   * @param message - User message text
   * @param schema - JSON Schema object the response must conform to
   * @param options - Generation options (grammarStr/grammarRoot will be set automatically)
   * @param cancelToken - Optional cancellation token
   * @returns Validated JSON object
   * @throws GenerationError if model output fails schema validation
   */
  async sendStructured(
    message: string,
    schema: Record<string, unknown>,
    options?: GenerateOptions,
    cancelToken?: CancelToken
  ): Promise<Record<string, unknown>> {
    // Generate GBNF grammar from the JSON schema
    const grammarStr = GbnfBuilder.fromJsonSchema(schema);

    // Merge grammar into generation options
    const grammarOptions: GenerateOptions = {
      ...options,
      grammarStr,
      grammarRoot: 'root',
    };

    // Delegate to send() for history management and rollback
    const responseText = await this.send(message, grammarOptions, cancelToken);

    // Parse JSON response
    let parsed: Record<string, unknown>;
    try {
      parsed = JSON.parse(responseText) as Record<string, unknown>;
    } catch {
      throw new GenerationError(
        'Model output is not valid JSON',
        `Output: "${responseText.length > 200 ? responseText.substring(0, 200) + '...' : responseText}"`
      );
    }

    // Validate against schema
    const validation = SchemaValidator.validate(parsed, schema);
    if (!validation.isValid) {
      throw new GenerationError(
        'Model output failed schema validation',
        `Errors: ${validation.errors.join(', ')}`
      );
    }

    return parsed;
  }

  // ---------------------------------------------------------------------------
  // Session management
  // ---------------------------------------------------------------------------

  /**
   * Resets the conversation, keeping only the system prompt.
   */
  async reset(): Promise<void> {
    const systemMsg = this.messages.find((m) => m.role === ChatRole.SYSTEM);
    this.messages = systemMsg ? [systemMsg] : [];
    await this.edgeVeda.resetContext();
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private formatPrompt(): string {
    return formatMessages(this.template, this.messages);
  }

  private _formatConversationWithTools(): string {
    if (!this._tools || this._tools.tools.length === 0) {
      return this.formatPrompt();
    }

    const toolSystemPrompt = ToolTemplate.formatToolSystemPrompt(
      this._toolFormat,
      this._tools.tools as ToolDefinition[],
      this.systemPromptText ?? undefined
    );

    const systemMsg: ChatMessage = {
      role: ChatRole.SYSTEM,
      content: toolSystemPrompt,
      timestamp: new Date(),
    };
    const nonSystemMessages = this.messages.filter((m) => m.role !== ChatRole.SYSTEM);
    return formatMessages(this.template, [systemMsg, ...nonSystemMessages]);
  }

  /**
   * Check if context window is getting full and summarize if needed.
   *
   * Triggers when estimated token usage exceeds 70% of available capacity.
   * Keeps the last 2 user turns and their replies intact; older messages
   * are summarized by the model. Falls back to simple truncation on error.
   */
  private async _summarizeIfNeeded(cancelToken?: CancelToken): Promise<void> {
    const formatted = this.formatPrompt();
    const estimatedTokens = Math.floor(formatted.length / 4);
    const availableTokens = this.maxContextLength - this.maxResponseTokens;

    if (estimatedTokens < availableTokens * 0.7) return;

    this._isSummarizing = true;
    try {
      // Find split point: keep last 2 user turns + their assistant replies
      let userCount = 0;
      let splitIndex = this.messages.length;
      for (let i = this.messages.length - 1; i >= 0; i--) {
        if (this.messages[i].role === ChatRole.USER) {
          userCount++;
        }
        if (userCount >= 2) {
          splitIndex = i;
          break;
        }
      }

      if (splitIndex <= 0) return;

      const oldMessages = this.messages.slice(0, splitIndex);
      const recentMessages = this.messages.slice(splitIndex);

      // Build summarization prompt
      let summaryPrompt = 'Summarize this conversation concisely. Keep key facts and decisions:\n';
      for (const msg of oldMessages) {
        if (msg.role === ChatRole.SUMMARY) {
          summaryPrompt += `summary: ${msg.content}\n`;
        } else {
          summaryPrompt += `${msg.role}: ${msg.content}\n`;
        }
      }

      cancelToken?.throwIfCancelled();

      // Generate summary with the model (low temperature for factual accuracy)
      const summaryText = await this.edgeVeda.generate(summaryPrompt, {
        maxTokens: 128,
        temperature: 0.3,
      });

      // Replace old messages with summary + recent messages
      this.messages = [
        {
          role: ChatRole.SUMMARY,
          content: summaryText,
          timestamp: new Date(),
        },
        ...recentMessages,
      ];
    } catch (_e) {
      // Fallback: simple truncation if summarization fails
      const targetTokens = Math.floor(
        (this.maxContextLength - this.maxResponseTokens) * 0.6
      );

      while (this.messages.length > 2) {
        const currentFormatted = this.formatPrompt();
        const currentTokens = Math.floor(currentFormatted.length / 4);
        if (currentTokens <= targetTokens) break;
        this.messages.splice(0, 1);
      }
    } finally {
      this._isSummarizing = false;
    }
  }

  private estimateTokens(): number {
    const totalChars = this.messages.reduce((sum, msg) => sum + msg.content.length, 0);
    return Math.floor(totalChars / 4);
  }

  private _deriveToolFormat(): ToolTemplateFormat {
    switch (this.template) {
      case ChatTemplate.QWEN3:
        return ToolTemplateFormat.QWEN3;
      case ChatTemplate.GEMMA3:
        return ToolTemplateFormat.GEMMA3;
      default:
        return ToolTemplateFormat.GENERIC;
    }
  }
}
